### DISCUSSION TITLE
Seal skills and Taker skills

### DISCUSSION THREAD

**Discussion by JesterWizard** [MokhaLeee/fe8u-cskillsys-kernel#241] [General]
These require bytes in RAM to store their values, but I'm not sure how to do it in this skill system. Shall I PR the skill info and let you decide on a working implementation @MokhaLeee or should I leave it for now?

**Comment by MokhaLeee** [comment]
SRAM EMS data is precious. Not only you need to consider whether the space is sufficient in **KERNEL**, you also need to reserve as much spaces as possible for users to customize their own **DEMO**. It is not reasonable to waste so much SRAM space to achieve just a few skills. These skills is planed to be abandoned from their original effects from the beginning of the skill system design,  and the bitmasked **stat-debuff** is introduced as an alternative.

**Comment by MokhaLeee** [comment]
But I think it is reasonable to modify their skill effects to make them to be a stat-debuff skill, such as make them just last only one turn.

**Reply by JesterWizard** [reply]
Mmm, maybe for the seal skills, but I want to see if I can find a want to make the taker skills last longer than that. Maybe do something like I did for speed boost and make it based on turn count, without using any additional space. We'll see.

**Comment by MokhaLeee** [comment]
I think I have found the idea: use 3 bits in stat-debuff as a turn duration but it is a bit cumbersome to implement. Maybe you can pull the PR on skill info first.

**Reply by JesterWizard** [reply]
Nice idea, I'll get on that then. This is for the seal skills?

**Reply by MokhaLeee** [reply]
I think both of them are okay

**Reply by JesterWizard** [reply]
If the plan is to use a turn count for taker skills then storing anything isn't necessary. That would only be required if you were planning to go with the original implementation and make it based on number of kills that chapter.

**Reply by MokhaLeee** [reply]
What means taker? amount of killing enemies? That's not okay lol

**Reply by JesterWizard** [reply]
Taker Defense for example `After attacking and defeating an enemy, gain +4 the relevant stat (Max +10)`

Obviously based on what you've said this isn't sustainable without a custom build. So I'll just revert to using turn count.

**Reply by JesterWizard** [reply]
I was thinking, it could be possible to do the taker skills as they were intended if BWL data was reset at the end of each chapter. Maybe have it as a configuration option?

**Reply by MokhaLeee** [reply]
What matters is not how to realise it, but whether it is worth introducing. The function is very simple to realise. You only need to malloc for a large enough space and register it in the save-data, and write a tick function on pre-phase and then everything is done.  The question is you need to malloc space for every unit for just this single skill. RAM space is very limited resource. I”m not quite agreeable since these skills occupy far more resources than others.

**Reply by JesterWizard** [reply]
But that's my point. If you have a configuration option to reset BWL data at the end of every chapter, you can use those pre-existing bytes for new skills without having to allocate more space.

**Reply by MokhaLeee** [reply]
bwl is not designed for every units. only part of characters has bwl

**Reply by MokhaLeee** [reply]
it is addressed by pid rather than  uid

**Comment by Veslyquix** [comment]
I liked the way the old skillsys did it: you could allocate 8 or so bytes per unit to be used for debuffs / buffs that deplete each turn, along with miscellaneous effects that wear off after a turn. Users could then customize this to their liking. I think you could offer a config option for these skills if you really hate using suspend sram. Typically users only want 200 convoy, debuffs/buffs, and 62 player units for ems usage.

**Comment by MokhaLeee** [comment]
>  allocate 8 or so bytes per unit to be used for debuffs / buffs
> 62 player units

There is no enough space to save these data :(

On the other hand, I tend to preserve more SRAM space for users, rather than kernel takes up all resources, so that we can see that users can do more unique modification.

https://github.com/MokhaLeee/fe8u-cskillsys-kernel/blob/main/Wizardry/Common/SaveData/Source/ModularSaveCore.c#L342

I left all remaining space to an external loader/saver, users can directly put their unique save/load function pointer to a [fixed position](https://github.com/MokhaLeee/fe8u-cskillsys-kernel/blob/main/Data/Reloc.event#L36) so that they can get access to unused SRAM.

> users want 200 convoy

This is also my opinion on convoy, since the relevant modifications have nothing to do with the skill system apart from SRAM allocation, so what we really need to do is to offer the hooks in save-data for customization, and then write doc. If developer need to expand them, I think maybe he should make it by himself.

**Reply by MokhaLeee** [reply]
Of course, it is also okay to directly integrate convoy expansion to kernel. After all, they are just very small functions.


It’s just that I think they are the same as [StatScreenfx](https://github.com/MokhaLeee/fe8u-cskillsys-kernel/tree/main/Wizardry/External/StatScreenfx) or [DrawMapActionAnimation](https://github.com/MokhaLeee/fe8u-cskillsys-kernel/tree/main/Wizardry/External/DrawMapActionAnimation), which should be removed someday in the future,
